pyclient/pydeephaven/table.py

```diff
@@ -0,0 +1,683 @@
+        returns the result table.
+
+        Args:
+            table (Table): the right-table of the join
+            keys (List[str]): the columns to match, can be a common name or an equal expression,
+                i.e. "col_a = col_b" for different column names
+            columns_to_add (List[str], optional): a list of the columns to be added from the right table to the result
+                table, can be renaming expressions, i.e. "new_col = col"; default is empty
+
+        Returns:
+            a Table object
+
+        Raises:
+            DHError
+
+        """
+        return self.session.table_service.nel_join(self, table, "exact", columns_to_match=keys,
+                                                   columns_to_add=columns_to_add)
+
+    def left_join(self, table: Table, keys: List[str], columns_to_add: List[str] = []) -> Table:
+        """ Perform a left-join between this table as the left table and another table as the right table) and
+        returns the result table.
+
+        Args:
+            table (Table): the right-table of the join
+            keys (List[str]): the columns to match, can be a common name or an equal expression,
+                i.e. "col_a = col_b" for different column names
+            columns_to_add (List[str], optional): a list of the columns to be added from the right table to the result
+                table, can be renaming expressions, i.e. "new_col = col"; default is empty
+
+        Returns:
+            a Table object
+
+        Raises:
+            DHError
+
+        """
+        return self.session.table_service.nel_join(self, table, "left", columns_to_match=keys,
+                                                   columns_to_add=columns_to_add)
+
+    def join(self, table: Table, keys: List[str] = [], columns_to_add: List[str] = [], reserve_bits: int = 10) -> Table:
+        """ Perform a cross-join between this table as the left table and another table as the right table) and
+        returns the result table.
+
+        Args:
+            table (Table): the right-table of the join
+            keys (List[str]): the columns to match, can be a common name or an equal expression,
+                i.e. "col_a = col_b" for different column names
+            columns_to_add (List[str], optional): a list of the columns to be added from the right table to the result
+                table, can be renaming expressions, i.e. "new_col = col"; default is empty
+            reserve_bits(int, optional): the number of bits of key-space to initially reserve per group; default is 10
+
+        Returns:
+            a Table object
+
+        Raises:
+            DHError
+
+        """
+        return self.session.table_service.cross_join(self, table, columns_to_match=keys,
+                                                     columns_to_add=columns_to_add, reserve_bits=reserve_bits)
+
+    def aj(self, table: Table, keys: List[str], columns_to_add: List[str] = []) -> Table:
```

Comment:
I think we need to expose aj match rules (LT, LTE) as an argument, default to LTE.